substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,785 @@
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+	/// Error happened while serializing the module.
+	Serialization,
+
+	/// Error happened while deserializing the module.
+	Deserialization,
+
+	/// Internal memory declaration has been found in the module.
+	InternalMemoryDeclared,
+
+	/// Gas instrumentation failed.
+	///
+	/// This most likely indicates the module isn't valid.
+	GasInstrumentation,
+
+	/// Stack instrumentation failed.
+	///
+	/// This  most likely indicates the module isn't valid.
+	StackHeightInstrumentation,
+
+	/// Error happened during invocation of the contract's entrypoint.
+	///
+	/// Most likely because of trap.
+	Invoke,
+
+	/// Error happened during instantiation.
+	///
+	/// This might indicate that `start` function trapped, or module isn't
+	/// instantiable and/or unlinkable.
+	Instantiate,
+
+	/// Memory creation error.
+	///
+	/// This might happen when the memory import has invalid descriptor or
+	/// requested too much resources.
+	Memory,
+}
+
+/// Enumerates all possible *special* trap conditions.
+///
+/// In this runtime traps used not only for signaling about errors but also
+/// to just terminate quickly in some cases.
+enum SpecialTrap {
+	/// Signals that trap was generated in response to call `ext_return` host function.
+	Return(Vec<u8>),
+}
+
+struct Runtime<'a, T: Trait + 'a, E: Ext<T> + 'a> {
+	ext: &'a mut E,
+	config: &'a Config<T>,
+	memory: sandbox::Memory,
+	gas_meter: &'a mut GasMeter<T>,
+	special_trap: Option<SpecialTrap>,
+}
+impl<'a, T: Trait, E: Ext<T> + 'a> Runtime<'a, T, E> {
+	fn memory(&self) -> &sandbox::Memory {
+		&self.memory
+	}
+	/// Save a data buffer as a result of the execution.
+	///
+	/// This function also charges gas for the returning.
+	///
+	/// Returns `Err` if there is not enough gas.
+	fn store_return_data(&mut self, data: Vec<u8>) -> Result<(), ()> {
+		let data_len = <T::Gas as As<u64>>::sa(data.len() as u64);
+		let price = (self.config.return_data_per_byte_cost)
+			.checked_mul(&data_len)
+			.ok_or_else(|| ())?;
+
+		match self.gas_meter.charge(price) {
+			GasMeterResult::Proceed => {
+				self.special_trap = Some(SpecialTrap::Return(data));
+				Ok(())
+			}
+			GasMeterResult::OutOfGas => Err(()),
+		}
+	}
+}
+
+fn to_execution_result<T: Trait, E: Ext<T>>(
+	runtime: Runtime<T, E>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let mut return_data = Vec::new();
+
+	// Check the exact type of the error. It could be plain trap or
+	// special runtime trap the we must recognize.
+	match (run_err, runtime.special_trap) {
+		// No traps were generated. Proceed normally.
+		(None, None) => {}
+		// Special case. The trap was the result of the execution `return` host function.
+		(Some(sandbox::Error::Execution), Some(SpecialTrap::Return(rd))) => return_data = rd,
+		// Any other kind of a trap should result in a failure.
+		(Some(_), _) => {
+			return Err(Error::Invoke);
```

Comment:
no need for a block, please make this a one-liner.

Reply:
ok!